gce/survival-training/dummy/train.py

```diff
@@ -0,0 +1,219 @@
+
+  last_path, last_index = latest_checkpoint(get_checkpoints(job_dir))
+  if last_index is not None:
+    current_checkpoint_index = last_index + 1
+    last_data = load_checkpoint(last_path)
+    last_hp = last_data.get("hyperparameters")
+    for hyperparameter in current_hyperparameters:
+      if (current_hyperparameters[hyperparameter] is not None and
+          current_hyperparameters[hyperparameter] != last_hp[hyperparameter]):
+        raise ValueError(
+            "Inconsistent values for {}: ".format(hyperparameter) +
+            "command line -- {}, checkpoint -- {}".format(
+                hyperparameters[hyperparameter],
+                last_data[hyperparameter]
+            )
+        )
+
+    current_hyperparameters = last_hp
+
+  train_step = trainer_initializer(hyperparameters)
+
+  def finished(step):
+    """Returns True if job is complete and False otherwise."""
+    if train_steps is None:
+      return False
+    else:
+      return step > train_steps
+
+  result = None
+  # TODO(nkashy1): Add test for "up to N steps" rather than "additional N steps"
+  i = current_checkpoint_index*checkpoint_steps + 1
+  while not finished(i):
+    result = train_step()
+
+    if i%checkpoint_steps == 0:
+      checkpoint_data = generate_checkpoint(
+          current_checkpoint_index,
+          hyperparameters,
+          result
+      )
+      save_checkpoint(job_dir, current_checkpoint_index, checkpoint_data)
+      current_checkpoint_index += 1
+
+    i += 1
+
+  checkpoint_data = generate_checkpoint(
+      current_checkpoint_index,
+      hyperparameters,
+      result
+  )
+  save_checkpoint(job_dir, current_checkpoint_index, checkpoint_data)
+
+
+def generate_checkpoint(step, hyperparameters, model_state):
+  """Generates checkpoint contents.
+
+  Args:
+    step: Training step at which this checkpoint was generated.
+    hyperparameters: Dictionary specifying the model hyperparameters.
+    model_state: A JSON serializable representation of the model state.
+
+  Returns:
+    Dictionary representing the content to be checkpointed.
+  """
+  checkpoint_data = {
+      "steps": step,
+      "hyperparameters": hyperparameters,
+      "model": model_state
+  }
+  return checkpoint_data
+
+
+def get_checkpoints(job_dir):
+  checkpoint_glob = os.path.join(job_dir, "dummy-checkpoint-*.json")
+  checkpoints = glob.glob(checkpoint_glob)
+  return checkpoints
+
+
+def latest_checkpoint(checkpoint_paths):
+  if not checkpoint_paths:
+    return (None, None)
+
+  checkpoint_indices = map(checkpoint_index, checkpoint_paths)
+  indexed_checkpoints = zip(checkpoint_paths, checkpoint_indices)
+  sorted_indexed_checkpoints = sorted(indexed_checkpoints, key=lambda p: p[1])
+  return sorted_indexed_checkpoints[-1]
+
+
+def checkpoint_index(checkpoint_path):
+  checkpoint_file = os.path.basename(checkpoint_path)
+  prefix = "dummy-checkpoint-"
+  suffix = ".json"
+  return int(checkpoint_file[len(prefix):-len(suffix)])
```

Comment:
What is this line doing?

Reply:
It's returning a numerical checkpoint number that we can do arithmetic with. It does look unnecessarily complicated. Would welcome ideas for improving it.

Reply:
(And also fragile.)

Reply:
using hash? does that suffice. Something like:

import hashlib
abs(hash(checkpoint_path)) & (10 ** 8)

Reply:
We want successive checkpoints to get successive numbers. That kind of hash does not achieve this purpose. We want the same number that got stored in the checkpoint path - `dummy-checkpoint-<n>.json` should correspond to the number n.

Reply:
how about use `re`?

Reply:
Will probably have to do that when we actually use the code (with sklearn model). For now, this code is just there for tests and to serve as an illustration of the checkpointing semantics. The tf estimator does its own checkpointing.

Would it be alright with you guys if we shelve this matter for now?

Reply:
This code has a bug. Last line is indexing the string starting at length(prefix) which can give you empty '' string. Also  the int(..) call won't work on string.